### ISSUE TITLE
MaP always 0. Using Yolov8

### ISSUE THREAD

**Issue by tamaratoma** [keras-team/keras#18935]
I am trying to train the https://public.roboflow.com/object-detection/self-driving-car/3 dataset with YOLOv8 using Keras (I used the example from [https://keras.io/examples/vision/yolov8/[Keras](https://keras.io/examples/vision/yolov8/%5BKeras) and Yolov8](https://keras.io/examples/vision/yolov8/)):

    """# Setup"""

    import os
    from tqdm.auto import tqdm
    import xml.etree.ElementTree as ET

    import tensorflow as tf
    from tensorflow import keras
    import requests
    import zipfile

    import keras_cv
    from keras_cv import bounding_box
    from keras_cv import visualization
    import cv2
    import numpy as np

    # Download dataset.
    def download_file(url, save_name):
        if not os.path.exists(save_name):
            print(f"Downloading file")
            file = requests.get(url, stream=True)
            total_size = int(file.headers.get('content-length', 0))
            block_size = 1024
            progress_bar = tqdm(
                total=total_size,
                unit='iB',
                unit_scale=True
            )
            with open(os.path.join(save_name), 'wb') as f:
                for data in file.iter_content(block_size):
                    progress_bar.update(len(data))
                    f.write(data)
            progress_bar.close()
        else:
            print('File already present')

    download_file(
        'https://app.roboflow.com/...',
        'cars.zip'
    )

    # Unzip the data file
    def unzip(zip_file=None):
        try:
            with zipfile.ZipFile(zip_file) as z:
                z.extractall("./")
                print("Extracted all")
        except:
            print("Invalid file")

    unzip('cars.zip')

    """# Hyperparameters"""

    SPLIT_RATIO = 0.2
    BATCH_SIZE = 4
    LEARNING_RATE = 0.001
    EPOCH = 5
    GLOBAL_CLIPNORM = 10.0

    # check if an image is grayscale
    def is_grayscale(image_path):
        img = cv2.imread(image_path)
        if len(img.shape) < 3: return True
        if img.shape[2]  == 1: return True
        b,g,r = img[:,:,0], img[:,:,1], img[:,:,2]
        if (b==g).all() and (b==r).all(): return True
        return False

    class_ids = [
        "car",
        "pedestrian",
        "trafficLight-Red",
        "trafficLight-Gree",
        "truck",
        "trafficLight",
        "biker",
        "trafficLight-RedLeft",
        "trafficLight-GreenLeft",
        "trafficLight-Yellow",
        "trafficLight-YellowLeft"
    ]
    class_mapping = dict(zip(range(len(class_ids)), class_ids))

    # Path to images and annotations
    path_images = "../content/train/"
    path_annot = "../content/train/"

    # Get all XML file paths in path_annot and sort them
    xml_files = sorted(
        [
            os.path.join(path_annot, file_name)
            for file_name in os.listdir(path_annot)
            if file_name.endswith(".xml")
        ]
    )

    # Get all JPEG image file paths in path_images and sort them
    jpg_files = sorted(
        [
            os.path.join(path_images, file_name)
            for file_name in os.listdir(path_images)
            if file_name.endswith(".jpg") and not is_grayscale(os.path.join(path_images,     file_name))
        ]
    )

    def parse_annotation(xml_file):
        tree = ET.parse(xml_file)
        root = tree.getroot()

        image_name = root.find("filename").text
        image_path = os.path.join(path_images, image_name)

        boxes = []
        classes = []
        for obj in root.iter("object"):
            cls = obj.find("name").text
            classes.append(cls)

            bbox = obj.find("bndbox")
            xmin = float(bbox.find("xmin").text)
            ymin = float(bbox.find("ymin").text)
            xmax = float(bbox.find("xmax").text)
            ymax = float(bbox.find("ymax").text)
            boxes.append([xmin, ymin, xmax, ymax])

        class_ids = [
            list(class_mapping.keys())[list(class_mapping.values()).index(cls)]
            for cls in classes
        ]
        return image_path, boxes, class_ids

    image_paths = []
    bbox = []
    classes = []
    for xml_file in tqdm(xml_files):
        image_path, boxes, class_ids = parse_annotation(xml_file)
        image_paths.append(image_path)
        bbox.append(boxes)
        classes.append(class_ids)

    bbox = tf.ragged.constant(bbox)
    classes = tf.ragged.constant(classes)
    image_paths = tf.ragged.constant(image_paths)

    data = tf.data.Dataset.from_tensor_slices((image_paths, classes, bbox))

    # Determine the number of validation samples
    num_val = int(len(xml_files) * SPLIT_RATIO)

    # Split the dataset into train and validation sets
    val_data = data.take(num_val)
    train_data = data.skip(num_val)

    def load_image(image_path):
        image = tf.io.read_file(image_path)
        image = tf.image.decode_jpeg(image, channels=3)
        return image

    def load_dataset(image_path, classes, bbox):
        # Read Image
        image = load_image(image_path)
        bounding_boxes = {
            "classes": tf.cast(classes, dtype=tf.float32),
            "boxes": bbox,
        }
        return {"images": tf.cast(image, tf.float32), "bounding_boxes": bounding_boxes}

    """# Data Augmentation"""

    augmenter = keras.Sequential(
        layers=[
            keras_cv.layers.RandomFlip(mode="horizontal", bounding_box_format="xyxy"),
            keras_cv.layers.RandomShear(
                x_factor=0.2, y_factor=0.2, bounding_box_format="xyxy"
            ),
            keras_cv.layers.JitteredResize(
                target_size=(640, 640), scale_factor=(0.75, 1.3), bounding_box_format="xyxy"
            ),
        ]
    )

    """# Creating Training Dataset"""

    train_ds = train_data.map(load_dataset, num_parallel_calls=tf.data.AUTOTUNE)
    train_ds = train_ds.shuffle(BATCH_SIZE * 4)
    train_ds = train_ds.ragged_batch(BATCH_SIZE, drop_remainder=True)
    train_ds = train_ds.map(augmenter, num_parallel_calls=tf.data.AUTOTUNE)

    """# Creating Validation Dataset"""

    resizing = keras_cv.layers.JitteredResize(
        target_size=(640, 640),
        scale_factor=(0.75, 1.3),
        bounding_box_format="xyxy",
    )

    val_ds = val_data.map(load_dataset, num_parallel_calls=tf.data.AUTOTUNE)
    val_ds = val_ds.shuffle(BATCH_SIZE * 4)
    val_ds = val_ds.ragged_batch(BATCH_SIZE, drop_remainder=True)
    val_ds = val_ds.map(resizing, num_parallel_calls=tf.data.AUTOTUNE)

    def dict_to_tuple(inputs):
        return inputs["images"], inputs["bounding_boxes"]

    train_ds = train_ds.map(dict_to_tuple, num_parallel_calls=tf.data.AUTOTUNE)
    train_ds = train_ds.prefetch(tf.data.AUTOTUNE)

    val_ds = val_ds.map(dict_to_tuple, num_parallel_calls=tf.data.AUTOTUNE)
    val_ds = val_ds.prefetch(tf.data.AUTOTUNE)
    
    """# Creating Model"""
    
    backbone = keras_cv.models.YOLOV8Backbone.from_preset(
        "yolo_v8_s_backbone_coco"  # We will use yolov8 small backbone with coco weights
    )
    
    yolo = keras_cv.models.YOLOV8Detector(
        num_classes=len(class_mapping),
        bounding_box_format="xyxy",
        backbone=backbone,
        fpn_depth=1,
    )

    """# Compile the Model"""
    
    optimizer = tf.keras.optimizers.Adam(
        learning_rate=LEARNING_RATE,
        global_clipnorm=GLOBAL_CLIPNORM,
    )

    yolo.compile(
        optimizer=optimizer, classification_loss="binary_crossentropy", box_loss="ciou"
    )

    """# COCO Metric Callback"""

    class EvaluateCOCOMetricsCallback(keras.callbacks.Callback):
        def __init__(self, data, save_path):
            super().__init__()
            self.data = data
            self.metrics = keras_cv.metrics.BoxCOCOMetrics(
                bounding_box_format="xyxy",
                evaluate_freq=1e9,
            )
    
            self.save_path = save_path
            self.best_map = -1.0

        def on_epoch_end(self, epoch, logs):
            self.metrics.reset_state()
            for batch in self.data:
                images, y_true = batch[0], batch[1]
                y_pred = self.model.predict(images, verbose=0)
                self.metrics.update_state(y_true, y_pred)

            metrics = self.metrics.result(force=True)
            logs.update(metrics)

            current_map = metrics["MaP"]
            if current_map > self.best_map:
                self.best_map = current_map
                self.model.save(self.save_path)  # Save the model when mAP improves

            return logs

    """# Train the Model"""

    yolo.fit(
        train_ds,
        validation_data=val_ds,
        epochs=3,
        callbacks=[EvaluateCOCOMetricsCallback(val_ds, "model.h5")],
    )

After 1 and many Epochs, I have a 
`...MaP: 0.0000e+00 - MaP@[IoU=50]: 0.0000e+00 - MaP@[IoU=75]: 0.0000e+00 - MaP@[area=small]: 0.0000e+00 - MaP@[area=medium]: 0.0000e+00 - MaP@[area=large]: 0.0000e+00.`

But when I train the EXACT SAME dataset with only using Yolov8, the mAP is not 0 and goes up with the time.
Why is that happening? Am I doing something wrong? It's not just with this dataset but with many others.

**Comment by hpsauce82** [comment]
Can confirm same issue when training the example code from keras-io/guides/keras_cv/object_detection_keras_cv.py 

Is there any update on this issue?

**Comment by alhaal** [comment]
Any updates on this issue? I have the same problem.

**Comment by dhantule** [comment]
Hi @tamaratoma, thanks for reporting this.

Could you please provide a reproducible colab to better investigate this issue ?
